pkg/cioutil/container_io.go

```diff
@@ -0,0 +1,221 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package cioutil
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+	"os"
+	"os/exec"
+	"runtime"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/containerd/containerd/cio"
+	"github.com/containerd/containerd/defaults"
+	"github.com/containerd/containerd/pkg/process"
+)
+
+const binaryIOProcTermTimeout = 12 * time.Second // Give logger process 10 seconds for cleanup
+
+// ncio is a basic container IO implementation.
+type ncio struct {
+	cmd     *exec.Cmd
+	config  cio.Config
+	wg      *sync.WaitGroup
+	closers []io.Closer
+	cancel  context.CancelFunc
+}
+
+var bufPool = sync.Pool{
+	New: func() interface{} {
+		buffer := make([]byte, 32<<10)
+		return &buffer
+	},
+}
+
+func (c *ncio) Config() cio.Config {
+	return c.config
+}
+
+func (c *ncio) Wait() {
+	if c.wg != nil {
+		c.wg.Wait()
+	}
+}
+
+func (c *ncio) Close() error {
+
+	var lastErr error
+
+	if c.cmd != nil && c.cmd.Process != nil {
+
+		// Send SIGTERM first, so logger process has a chance to flush and exit properly
+		if err := c.cmd.Process.Signal(syscall.SIGTERM); err != nil {
+			lastErr = fmt.Errorf("failed to send SIGTERM: %w", err)
```

Comment:
This takes time and causes:
- https://github.com/containerd/nerdctl/issues/2726

Reply:
Thanks for fixing the regression in #2727 , resolving this thread.